File: packages/cli/src/monorepo.ts
import type { Dirent } from "node:fs";
import { cp, mkdir, readdir, rename, rm, stat } from "node:fs/promises";
import { join, resolve } from "node:path";

/**
 * Monorepo task engine backing the `monorepo:run` command. It provides the
 * pieces the Talos workspace relies on — project discovery, a workspace
 * dependency graph and content-addressed task caching — with a granular
 * hash: every non-artifact file of a target and of its transitive workspace
 * dependencies is content-hashed individually, together with the script text,
 * the root config files and a cache version. Cache entries live under
 * `var/cache/monorepo/<hash>/` and store the task's captured output plus its
 * output artifacts (`dist` by default) so hits replay logs and restore files.
 */

export type MonorepoTargetTypeName = "package" | "module";

export type MonorepoTargetType = {
  /** Unique key, e.g. `packages/billing`. */
  key: string;
  /** Directory name, e.g. `billing`. */
  name: string;
  type: MonorepoTargetTypeName;
  /** Absolute directory of the target. */
  dir: string;
  scripts: Record<string, string>;
  /** Keys of other discovered targets this one depends on. */
  workspaceDeps: string[];
  /** Directories (relative to the target) captured and restored by the cache. */
  outputs: string[];
};

export type CacheEntryMetaType = {
  version: number;
  target: string;
  command: string;
  hash: string;
  createdAt: string;
  durationMs: number;
  outputs: string[];
};

export const MONOREPO_CACHE_VERSION = 1;
export const MONOREPO_CACHE_DIR = join("var", "cache", "monorepo");

const TARGET_ROOTS: { dirName: string; type: MonorepoTargetTypeName }[] = [
  { dirName: "packages", type: "package" },
  { dirName: "modules", type: "module" },
];

// Build artifacts and dependency folders never participate in the input hash.
const EXCLUDED_DIRS = new Set(["node_modules", "dist", "var", "coverage", ".git", ".temp", ".turbo"]);

// Root-level files whose content invalidates every task when they change.
const ROOT_INPUT_FILES = ["package.json", "bun.lock", "tsconfig.json", "biome.jsonc"];

const DEFAULT_OUTPUTS = ["dist"];

type PackageJsonType = {
  name?: string;
  scripts?: Record<string, string>;
  dependencies?: Record<string, string>;
  devDependencies?: Record<string, string>;
  peerDependencies?: Record<string, string>;
  talos?: { monorepo?: { outputs?: string[] } };
};

/**
 * Discover every package and module of the workspace rooted at `rootDir` that
 * has a package.json, and resolve their dependencies on one another (any
 * declared dependency whose name matches another discovered target).
 */
export const discoverTargets = async (rootDir: string): Promise<MonorepoTargetType[]> => {
  const targets: MonorepoTargetType[] = [];
  const keyByPackageName = new Map<string, string>();
  const declaredDeps = new Map<string, string[]>();

  for (const { dirName, type } of TARGET_ROOTS) {
    let entries: Dirent[];
    try {
      entries = await readdir(join(rootDir, dirName), { withFileTypes: true });
    } catch {
      continue;
    }

    for (const entry of entries) {
      if (!entry.isDirectory()) continue;

      const dir = join(rootDir, dirName, entry.name);
      const packageJsonFile = Bun.file(join(dir, "package.json"));
      if (!(await packageJsonFile.exists())) continue;

      const packageJson: PackageJsonType = await packageJsonFile.json();
      const key = `${dirName}/${entry.name}`;

      if (packageJson.name) keyByPackageName.set(packageJson.name, key);
      declaredDeps.set(key, [
        ...Object.keys(packageJson.dependencies ?? {}),
        ...Object.keys(packageJson.devDependencies ?? {}),
        ...Object.keys(packageJson.peerDependencies ?? {}),
      ]);

      targets.push({
        key,
        name: entry.name,
        type,
        dir,
        scripts: packageJson.scripts ?? {},
        workspaceDeps: [],
        outputs: packageJson.talos?.monorepo?.outputs ?? DEFAULT_OUTPUTS,
      });
    }
  }

  for (const target of targets) {
    target.workspaceDeps = (declaredDeps.get(target.key) ?? [])
      .map((name) => keyByPackageName.get(name))
      .filter((key): key is string => key !== undefined && key !== target.key);
  }

  return targets;
};

/**
 * Order targets so every target comes after its workspace dependencies, so a
 * target's `build` runs only once its dependencies have built. Cycles are
 * tolerated: members of a cycle keep their discovery order.
 */
export const sortTargetsByDependencies = (targets: MonorepoTargetType[]): MonorepoTargetType[] => {
  const byKey = new Map(targets.map((target) => [target.key, target]));
  const visited = new Set<string>();
  const visiting = new Set<string>();
  const sorted: MonorepoTargetType[] = [];

  const visit = (target: MonorepoTargetType): void => {
    if (visited.has(target.key) || visiting.has(target.key)) return;
    visiting.add(target.key);
    for (const depKey of target.workspaceDeps) {
      const dep = byKey.get(depKey);
      if (dep) visit(dep);
    }
    visiting.delete(target.key);
    visited.add(target.key);
    sorted.push(target);
  };

  for (const target of targets) visit(target);

  return sorted;
};

// List every hashable file of a directory, recursively, as sorted relative paths.
const collectFiles = async (dir: string, base = ""): Promise<string[]> => {
  let entries: Dirent[];
  try {
    entries = await readdir(dir, { withFileTypes: true });
  } catch {
    return [];
  }

  const files: string[] = [];
  for (const entry of entries) {
    const relPath = base ? `${base}/${entry.name}` : entry.name;
    if (entry.isDirectory()) {
      if (!EXCLUDED_DIRS.has(entry.name)) files.push(...(await collectFiles(join(dir, entry.name), relPath)));
    } else if (entry.isFile()) {
      files.push(relPath);
    }
  }

  return files.sort();
};

/**
 * True when `rootDir` is itself the top level of a git repository. Only then
 * can `git ls-files` be trusted for input hashing: in a nested, repo-less
 * folder the surrounding repository's ignore rules could hide every file.
 */
export const isGitWorkspaceRoot = async (rootDir: string): Promise<boolean> => {
  try {
    const proc = Bun.spawn(["git", "rev-parse", "--show-toplevel"], {
      cwd: rootDir,
      stdout: "pipe",
      stderr: "ignore",
    });
    const [stdout, exitCode] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
    if (exitCode !== 0) return false;
    return resolve(stdout.trim()) === resolve(rootDir);
  } catch {
    return false;
  }
};

// List a target's files through git: tracked files plus untracked ones that are
// not ignored. This keeps git-ignored artifacts (temp dirs written by tests,
// local scratch files) out of the fingerprint. Returns null when git fails.
const collectFilesWithGit = async (dir: string): Promise<string[] | null> => {
  try {
    const proc = Bun.spawn(["git", "ls-files", "--cached", "--others", "--exclude-standard", "-z"], {
      cwd: dir,
      stdout: "pipe",
      stderr: "ignore",
    });
    const [stdout, exitCode] = await Promise.all([new Response(proc.stdout).text(), proc.exited]);
    if (exitCode !== 0) return null;

    const files = stdout
      .split("\0")
      .filter(Boolean)
      .filter((file) => !file.split("/").some((segment) => EXCLUDED_DIRS.has(segment)));

    return [...new Set(files)].sort();
  } catch {
    return null;
  }
};

// Hash a file's content, or null when it cannot be read (e.g. tracked by git
// but deleted from the working tree — its manifest line is simply omitted).
const hashFile = async (path: string): Promise<string | null> => {
  try {
    const hasher = new Bun.CryptoHasher("sha256");
    hasher.update(await Bun.file(path).arrayBuffer());
    return hasher.digest("hex");
  } catch {
    return null;
  }
};

/**
 * Content-hash every file of a directory into a single fingerprint. When the
 * workspace is a git repository, files are listed through git so ignored
 * artifacts never invalidate the fingerprint; otherwise the directory is walked
 * (build and dependency folders are always excluded).
 */
export const fingerprintDir = async (dir: string, useGit = false): Promise<string> => {
  const files = (useGit ? await collectFilesWithGit(dir) : null) ?? (await collectFiles(dir));
  const hasher = new Bun.CryptoHasher("sha256");
  for (const file of files) {
    const fileHash = await hashFile(join(dir, file));
    if (fileHash !== null) hasher.update(`${file}=${fileHash}\n`);
  }
  return hasher.digest("hex");
};

/**
 * Content-hash every file of a target into a single fingerprint. Fingerprints
 * are memoized per target key in `memo` so shared dependencies are hashed once
 * per run.
 */
export const fingerprintTarget = (
  target: MonorepoTargetType,
  memo: Map<string, Promise<string>>,
  useGit = false,
): Promise<string> => {
  const memoized = memo.get(target.key);
  if (memoized) return memoized;

  const promise = fingerprintDir(target.dir, useGit);
  memo.set(target.key, promise);
  return promise;
};

/** Fingerprint the root config files shared by every task of the workspace. */
export const hashRootInputs = async (rootDir: string): Promise<string> => {
  const hasher = new Bun.CryptoHasher("sha256");
  for (const name of ROOT_INPUT_FILES) {
    const fileHash = await hashFile(join(rootDir, name));
    if (fileHash !== null) hasher.update(`${name}=${fileHash}\n`);
  }
  return hasher.digest("hex");
};

// Every workspace dependency reachable from the target, direct or transitive.
const transitiveDeps = (target: MonorepoTargetType, byKey: Map<string, MonorepoTargetType>): MonorepoTargetType[] => {
  const seen = new Set<string>([target.key]);
  const queue = [...target.workspaceDeps];
  const deps: MonorepoTargetType[] = [];

  while (queue.length > 0) {
    const key = queue.shift() as string;
    if (seen.has(key)) continue;
    seen.add(key);
    const dep = byKey.get(key);
    if (!dep) continue;
    deps.push(dep);
    queue.push(...dep.workspaceDeps);
  }

  return deps;
};

/**
 * Compute the cache hash of one task (`target` × `command`). The hash covers
 * the cache version, the script text, the target's own file fingerprint, the
 * fingerprints of all transitive workspace dependencies and the root inputs —
 * so a task re-runs exactly when something it can observe has changed.
 */
export const computeTaskHash = async (
  target: MonorepoTargetType,
  command: string,
  targets: MonorepoTargetType[],
  rootHash: string,
  memo: Map<string, Promise<string>>,
  useGit = false,
): Promise<string> => {
  const byKey = new Map(targets.map((entry) => [entry.key, entry]));
  const deps = transitiveDeps(target, byKey);
  const depLines = await Promise.all(
    deps.map(async (dep) => `${dep.key}=${await fingerprintTarget(dep, memo, useGit)}`),
  );

  const hasher = new Bun.CryptoHasher("sha256");
  hasher.update(
    [
      `version=${MONOREPO_CACHE_VERSION}`,
      `target=${target.key}`,
      `command=${command}`,
      `script=${target.scripts[command] ?? ""}`,
      `root=${rootHash}`,
      `self=${await fingerprintTarget(target, memo, useGit)}`,
      ...depLines.sort(),
    ].join("\n"),
  );

  return hasher.digest("hex");
};

/** Read a cache entry, returning its metadata and replayable output, or null on miss. */
export const readCacheEntry = async (
  cacheDir: string,
  hash: string,
): Promise<{ meta: CacheEntryMetaType; output: string } | null> => {
  const metaFile = Bun.file(join(cacheDir, hash, "meta.json"));
  if (!(await metaFile.exists())) return null;

  try {
    const meta: CacheEntryMetaType = await metaFile.json();
    const outputFile = Bun.file(join(cacheDir, hash, "output.log"));
    const output = (await outputFile.exists()) ? await outputFile.text() : "";
    return { meta, output };
  } catch {
    return null;
  }
};

const pathExists = async (path: string): Promise<boolean> => {
  try {
    await stat(path);
    return true;
  } catch {
    return false;
  }
};

/** Copy the cached output artifacts of an entry back into the target directory. */
export const restoreCacheOutputs = async (
  cacheDir: string,
  meta: CacheEntryMetaType,
  targetDir: string,
): Promise<void> => {
  for (const output of meta.outputs) {
    const source = join(cacheDir, meta.hash, "outputs", output);
    if (!(await pathExists(source))) continue;
    await rm(join(targetDir, output), { recursive: true, force: true });
    await cp(source, join(targetDir, output), { recursive: true });
  }
};

/**
 * Persist a successful task run: metadata, replayable output and the output
 * artifacts that exist. The entry is assembled in a temp directory and renamed
 * into place so a crashed run never leaves a half-written entry behind.
 */
export const writeCacheEntry = async (
  cacheDir: string,
  meta: CacheEntryMetaType,
  output: string,
  targetDir: string,
): Promise<void> => {
  const tempDir = join(cacheDir, `.tmp-${meta.hash}`);
  const entryDir = join(cacheDir, meta.hash);

  await rm(tempDir, { recursive: true, force: true });
  await mkdir(tempDir, { recursive: true });

  const capturedOutputs: string[] = [];
  for (const out of meta.outputs) {
    const source = join(targetDir, out);
    if (!(await pathExists(source))) continue;
    await cp(source, join(tempDir, "outputs", out), { recursive: true });
    capturedOutputs.push(out);
  }

  await Bun.write(join(tempDir, "meta.json"), JSON.stringify({ ...meta, outputs: capturedOutputs }, null, 2));
  await Bun.write(join(tempDir, "output.log"), output);

  await rm(entryDir, { recursive: true, force: true });
  await rename(tempDir, entryDir);
};

File: packages/rust-cli/src/utils/monorepo.rs
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::time::UNIX_EPOCH;

use dashmap::DashMap;
use rayon::prelude::*;
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

pub const MONOREPO_CACHE_VERSION: u32 = 1;
pub const MONOREPO_CACHE_DIR: &str = "var/cache/monorepo";

const TARGET_ROOTS: &[(&str, TargetType)] = &[
    ("packages", TargetType::Package),
    ("modules", TargetType::Module),
];

const EXCLUDED_DIRS: &[&str] = &[
    "node_modules",
    "dist",
    "var",
    "coverage",
    ".git",
    ".temp",
    ".turbo",
];

const ROOT_INPUT_FILES: &[&str] = &["package.json", "bun.lock", "tsconfig.json", "biome.jsonc"];

const DEFAULT_OUTPUTS: &[&str] = &["dist"];

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum TargetType {
    Package,
    Module,
}

impl TargetType {
    pub fn as_str(&self) -> &'static str {
        match self {
            TargetType::Package => "package",
            TargetType::Module => "module",
        }
    }
}

#[derive(Clone, Debug)]
pub struct MonorepoTarget {
    pub key: String,
    pub name: String,
    pub target_type: TargetType,
    pub dir: PathBuf,
    pub scripts: HashMap<String, String>,
    pub workspace_deps: Vec<String>,
    pub outputs: Vec<String>,
}

#[derive(Deserialize)]
struct PackageJson {
    #[serde(default)]
    name: Option<String>,
    #[serde(default)]
    scripts: HashMap<String, String>,
    #[serde(default)]
    dependencies: HashMap<String, String>,
    #[serde(default)]
    #[serde(rename = "devDependencies")]
    dev_dependencies: HashMap<String, String>,
    #[serde(default)]
    #[serde(rename = "peerDependencies")]
    peer_dependencies: HashMap<String, String>,
    #[serde(default)]
    talos: Option<TalosField>,
}

#[derive(Deserialize)]
struct TalosField {
    #[serde(default)]
    monorepo: Option<TalosMonorepoField>,
}

#[derive(Deserialize)]
struct TalosMonorepoField {
    #[serde(default)]
    outputs: Option<Vec<String>>,
}

pub fn discover_targets(root_dir: &Path) -> Vec<MonorepoTarget> {
    let mut targets: Vec<MonorepoTarget> = Vec::new();
    let mut key_by_package_name: HashMap<String, String> = HashMap::new();
    let mut declared_deps: HashMap<String, Vec<String>> = HashMap::new();

    for (dir_name, target_type) in TARGET_ROOTS {
        let Ok(entries) = fs::read_dir(root_dir.join(dir_name)) else {
            continue;
        };
        let mut names: Vec<String> = entries
            .flatten()
            .filter(|e| e.path().is_dir())
            .filter_map(|e| e.file_name().into_string().ok())
            .collect();
        names.sort();

        for name in names {
            let dir = root_dir.join(dir_name).join(&name);
            let package_json_path = dir.join("package.json");
            let Ok(raw) = fs::read_to_string(&package_json_path) else {
                continue;
            };
            let Ok(package_json) = serde_json::from_str::<PackageJson>(&raw) else {
                continue;
            };

            let key = format!("{dir_name}/{name}");
            if let Some(pkg_name) = &package_json.name {
                key_by_package_name.insert(pkg_name.clone(), key.clone());
            }
            let mut deps: Vec<String> = package_json.dependencies.keys().cloned().collect();
            deps.extend(package_json.dev_dependencies.keys().cloned());
            deps.extend(package_json.peer_dependencies.keys().cloned());
            declared_deps.insert(key.clone(), deps);

            let outputs = package_json
                .talos
                .and_then(|t| t.monorepo)
                .and_then(|m| m.outputs)
                .unwrap_or_else(|| DEFAULT_OUTPUTS.iter().map(|s| s.to_string()).collect());

            targets.push(MonorepoTarget {
                key,
                name,
                target_type: *target_type,
                dir,
                scripts: package_json.scripts,
                workspace_deps: Vec::new(),
                outputs,
            });
        }
    }

    for target in &mut targets {
        target.workspace_deps = declared_deps
            .get(&target.key)
            .cloned()
            .unwrap_or_default()
            .iter()
            .filter_map(|name| key_by_package_name.get(name))
            .filter(|key| *key != &target.key)
            .cloned()
            .collect();
    }

    targets
}

pub fn sort_targets_by_dependencies(targets: &[MonorepoTarget]) -> Vec<MonorepoTarget> {
    let by_key: HashMap<&str, &MonorepoTarget> =
        targets.iter().map(|t| (t.key.as_str(), t)).collect();
    let mut visited: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut visiting: std::collections::HashSet<String> = std::collections::HashSet::new();
    let mut sorted: Vec<MonorepoTarget> = Vec::new();

    fn visit(
        target: &MonorepoTarget,
        by_key: &HashMap<&str, &MonorepoTarget>,
        visited: &mut std::collections::HashSet<String>,
        visiting: &mut std::collections::HashSet<String>,
        sorted: &mut Vec<MonorepoTarget>,
    ) {
        if visited.contains(&target.key) || visiting.contains(&target.key) {
            return;
        }
        visiting.insert(target.key.clone());
        for dep_key in &target.workspace_deps {
            if let Some(dep) = by_key.get(dep_key.as_str()) {
                visit(dep, by_key, visited, visiting, sorted);
            }
        }
        visiting.remove(&target.key);
        visited.insert(target.key.clone());
        sorted.push(target.clone());
    }

    for target in targets {
        visit(target, &by_key, &mut visited, &mut visiting, &mut sorted);
    }

    sorted
}

fn walk_files(dir: &Path, base: &str, files: &mut Vec<String>) {
    let Ok(entries) = fs::read_dir(dir) else {
        return;
    };
    for entry in entries.flatten() {
        let path = entry.path();
        let name = entry.file_name();
        let name_str = name.to_string_lossy();
        let rel_path = if base.is_empty() {
            name_str.to_string()
        } else {
            format!("{base}/{name_str}")
        };
        if path.is_dir() {
            if !EXCLUDED_DIRS.contains(&name_str.as_ref()) {
                walk_files(&path, &rel_path, files);
            }
        } else if path.is_file() {
            files.push(rel_path);
        }
    }
}

fn collect_files(dir: &Path) -> Vec<String> {
    let mut files = Vec::new();
    walk_files(dir, "", &mut files);
    files.sort();
    files
}

pub fn resolve_biome_command(start_dir: &Path) -> Vec<String> {
    resolve_local_bin(start_dir, "biome")
}

pub fn resolve_tsc_command(start_dir: &Path) -> Vec<String> {
    resolve_local_bin(start_dir, "tsc")
}

fn resolve_local_bin(start_dir: &Path, bin: &str) -> Vec<String> {
    let mut dir = start_dir.to_path_buf();
    loop {
        let candidate = dir.join(format!("node_modules/.bin/{bin}"));
        if candidate.is_file() {
            return vec![candidate.to_string_lossy().to_string()];
        }
        if !dir.pop() {
            break;
        }
    }
    vec!["bunx".to_string(), bin.to_string()]
}

pub fn is_git_workspace_root(root_dir: &Path) -> bool {
    let Some(toplevel) = crate::utils::git::toplevel(root_dir) else {
        return false;
    };
    let Ok(resolved_toplevel) = fs::canonicalize(&toplevel) else {
        return false;
    };
    let Ok(resolved_root) = fs::canonicalize(root_dir) else {
        return false;
    };
    resolved_toplevel == resolved_root
}

fn collect_files_with_git(dir: &Path) -> Option<Vec<String>> {
    let repo = git2::Repository::discover(dir).ok()?;
    let workdir = repo.workdir()?;
    let prefix = dir.strip_prefix(workdir).ok()?;
    let prefix_str = prefix.to_string_lossy().replace('\\', "/");
    let strip = |path: &str| -> Option<String> {
        if prefix_str.is_empty() {
            Some(path.to_string())
        } else {
            path.strip_prefix(&prefix_str)
                .map(|rest| rest.trim_start_matches('/').to_string())
        }
    };

    let mut files = Vec::new();

    let index = repo.index().ok()?;
    for entry in index.iter() {
        let path = String::from_utf8_lossy(&entry.path).replace('\\', "/");
        if let Some(relative) = strip(&path) {
            files.push(relative);
        }
    }

    let mut status_options = git2::StatusOptions::new();
    status_options
        .include_untracked(true)
        .recurse_untracked_dirs(true)
        .include_ignored(false)
        .include_unmodified(false);
    if !prefix_str.is_empty() {
        status_options.pathspec(&prefix_str);
    }
    let statuses = repo.statuses(Some(&mut status_options)).ok()?;
    for status_entry in statuses.iter() {
        if !status_entry.status().contains(git2::Status::WT_NEW) {
            continue;
        }
        let Some(path) = status_entry.path() else {
            continue;
        };
        if let Some(relative) = strip(path) {
            files.push(relative);
        }
    }

    let mut files: Vec<String> = files
        .into_iter()
        .filter(|file| {
            !file
                .split('/')
                .any(|segment| EXCLUDED_DIRS.contains(&segment))
        })
        .collect();
    files.sort();
    files.dedup();
    Some(files)
}

fn hash_file(path: &Path) -> Option<String> {
    let bytes = fs::read(path).ok()?;
    let mut hasher = Sha256::new();
    hasher.update(&bytes);
    Some(format!("{:x}", hasher.finalize()))
}

#[derive(Clone, Serialize, Deserialize)]
pub struct FileHashRecord {
    pub size: u64,
    #[serde(rename = "mtimeMs")]
    pub mtime_ms: f64,
    pub hash: String,
}

pub type FileHashCache = DashMap<String, FileHashRecord>;

pub type FingerprintMemo = DashMap<String, String>;

const FILEHASH_CACHE_FILE: &str = "filehashes.json";

pub fn load_file_hash_cache(cache_dir: &Path) -> FileHashCache {
    let path = cache_dir.join(FILEHASH_CACHE_FILE);
    let Ok(raw) = fs::read_to_string(&path) else {
        return FileHashCache::new();
    };
    serde_json::from_str(&raw).unwrap_or_default()
}

pub fn save_file_hash_cache(cache_dir: &Path, cache: &FileHashCache) {
    let _ = fs::create_dir_all(cache_dir);
    if let Ok(json) = serde_json::to_string(cache) {
        let _ = fs::write(cache_dir.join(FILEHASH_CACHE_FILE), json);
    }
}

fn mtime_millis(metadata: &fs::Metadata) -> f64 {
    metadata
        .modified()
        .ok()
        .and_then(|time| time.duration_since(UNIX_EPOCH).ok())
        .map(|duration| duration.as_secs_f64() * 1000.0)
        .unwrap_or(0.0)
}

fn hash_file_cached(path: &Path, cache: &FileHashCache) -> Option<String> {
    let metadata = fs::metadata(path).ok()?;
    let size = metadata.len();
    let mtime_ms = mtime_millis(&metadata);
    let path_key = path.to_string_lossy().to_string();

    if let Some(record) = cache.get(&path_key)
        && record.size == size
        && record.mtime_ms == mtime_ms
    {
        return Some(record.hash.clone());
    }

    let hash = hash_file(path)?;
    cache.insert(
        path_key,
        FileHashRecord {
            size,
            mtime_ms,
            hash: hash.clone(),
        },
    );
    Some(hash)
}

pub fn fingerprint_dir(dir: &Path, use_git: bool, file_hash_cache: &FileHashCache) -> String {
    let files = if use_git {
        collect_files_with_git(dir)
    } else {
        None
    }
    .unwrap_or_else(|| collect_files(dir));

    let hashes: Vec<Option<String>> = files
        .par_iter()
        .map(|file| hash_file_cached(&dir.join(file), file_hash_cache))
        .collect();

    let mut hasher = Sha256::new();
    for (file, hash) in files.iter().zip(hashes) {
        if let Some(hash) = hash {
            hasher.update(format!("{file}={hash}\n").as_bytes());
        }
    }
    format!("{:x}", hasher.finalize())
}

pub fn fingerprint_target(
    target: &MonorepoTarget,
    memo: &FingerprintMemo,
    use_git: bool,
    file_hash_cache: &FileHashCache,
) -> String {
    if let Some(cached) = memo.get(&target.key) {
        return cached.clone();
    }
    let fingerprint = fingerprint_dir(&target.dir, use_git, file_hash_cache);
    memo.insert(target.key.clone(), fingerprint.clone());
    fingerprint
}

pub fn hash_root_inputs(root_dir: &Path) -> String {
    let mut hasher = Sha256::new();
    for name in ROOT_INPUT_FILES {
        if let Some(hash) = hash_file(&root_dir.join(name)) {
            hasher.update(format!("{name}={hash}\n").as_bytes());
        }
    }
    format!("{:x}", hasher.finalize())
}

fn transitive_deps<'a>(
    target: &MonorepoTarget,
    by_key: &HashMap<&str, &'a MonorepoTarget>,
) -> Vec<&'a MonorepoTarget> {
    let mut seen: std::collections::HashSet<String> =
        std::collections::HashSet::from([target.key.clone()]);
    let mut queue: Vec<String> = target.workspace_deps.clone();
    let mut deps = Vec::new();

    while let Some(key) = queue.pop() {
        if seen.contains(&key) {
            continue;
        }
        seen.insert(key.clone());
        if let Some(dep) = by_key.get(key.as_str()) {
            deps.push(*dep);
            queue.extend(dep.workspace_deps.clone());
        }
    }

    deps
}

pub fn compute_task_hash(
    target: &MonorepoTarget,
    command: &str,
    targets: &[MonorepoTarget],
    root_hash: &str,
    memo: &FingerprintMemo,
    use_git: bool,
    file_hash_cache: &FileHashCache,
) -> String {
    let by_key: HashMap<&str, &MonorepoTarget> =
        targets.iter().map(|t| (t.key.as_str(), t)).collect();
    let deps = transitive_deps(target, &by_key);
    let mut dep_lines: Vec<String> = deps
        .par_iter()
        .map(|dep| {
            format!(
                "{}={}",
                dep.key,
                fingerprint_target(dep, memo, use_git, file_hash_cache)
            )
        })
        .collect();
    dep_lines.sort();

    let self_fingerprint = fingerprint_target(target, memo, use_git, file_hash_cache);
    let script = target.scripts.get(command).cloned().unwrap_or_default();

    let mut lines = vec![
        format!("version={MONOREPO_CACHE_VERSION}"),
        format!("target={}", target.key),
        format!("command={command}"),
        format!("script={script}"),
        format!("root={root_hash}"),
        format!("self={self_fingerprint}"),
    ];
    lines.extend(dep_lines);

    let mut hasher = Sha256::new();
    hasher.update(lines.join("\n").as_bytes());
    format!("{:x}", hasher.finalize())
}

#[derive(Clone, Serialize, Deserialize)]
pub struct CacheEntryMeta {
    pub version: u32,
    pub target: String,
    pub command: String,
    pub hash: String,
    #[serde(rename = "createdAt")]
    pub created_at: String,
    #[serde(rename = "durationMs")]
    pub duration_ms: u64,
    pub outputs: Vec<String>,
}

pub fn read_cache_entry(cache_dir: &Path, hash: &str) -> Option<CacheEntryMeta> {
    let meta_path = cache_dir.join(hash).join("meta.json");
    let raw_meta = fs::read_to_string(&meta_path).ok()?;
    serde_json::from_str(&raw_meta).ok()
}

pub fn restore_cache_outputs(cache_dir: &Path, meta: &CacheEntryMeta, target_dir: &Path) {
    for output in &meta.outputs {
        let source = cache_dir.join(&meta.hash).join("outputs").join(output);
        if !source.exists() {
            continue;
        }
        let dest = target_dir.join(output);
        let _ = fs::remove_dir_all(&dest);
        let _ = copy_dir_recursive(&source, &dest);
    }
}

fn copy_dir_recursive(source: &Path, dest: &Path) -> std::io::Result<()> {
    fs::create_dir_all(dest)?;
    for entry in fs::read_dir(source)? {
        let entry = entry?;
        let dest_path = dest.join(entry.file_name());
        if entry.path().is_dir() {
            copy_dir_recursive(&entry.path(), &dest_path)?;
        } else {
            fs::copy(entry.path(), &dest_path)?;
        }
    }
    Ok(())
}

pub fn write_cache_entry(cache_dir: &Path, meta: &CacheEntryMeta, output: &str, target_dir: &Path) {
    let temp_dir = cache_dir.join(format!(".tmp-{}", meta.hash));
    let entry_dir = cache_dir.join(&meta.hash);

    let _ = fs::remove_dir_all(&temp_dir);
    let _ = fs::create_dir_all(&temp_dir);

    let mut captured_outputs: Vec<String> = Vec::new();
    for out in &meta.outputs {
        let source = target_dir.join(out);
        if !source.exists() {
            continue;
        }
        if copy_dir_recursive(&source, &temp_dir.join("outputs").join(out)).is_ok() {
            captured_outputs.push(out.clone());
        }
    }

    let mut persisted_meta = meta.clone();
    persisted_meta.outputs = captured_outputs;
    if let Ok(json) = serde_json::to_string_pretty(&persisted_meta) {
        let _ = fs::write(temp_dir.join("meta.json"), json);
    }
    let _ = fs::write(temp_dir.join("output.log"), output);

    let _ = fs::remove_dir_all(&entry_dir);
    let _ = fs::rename(&temp_dir, &entry_dir);
}

#[cfg(test)]
mod git_backed_file_listing_tests {
    use super::*;
    use std::process::Command;

    fn git_ls_files(dir: &Path) -> Option<Vec<String>> {
        let output = Command::new("git")
            .args([
                "ls-files",
                "--cached",
                "--others",
                "--exclude-standard",
                "-z",
            ])
            .current_dir(dir)
            .output()
            .ok()?;
        if !output.status.success() {
            return None;
        }
        let stdout = String::from_utf8_lossy(&output.stdout);
        let mut files: Vec<String> = stdout
            .split('\0')
            .filter(|s| !s.is_empty())
            .filter(|file| {
                !file
                    .split('/')
                    .any(|segment| EXCLUDED_DIRS.contains(&segment))
            })
            .map(|s| s.to_string())
            .collect();
        files.sort();
        files.dedup();
        Some(files)
    }

    #[test]
    fn collect_files_with_git_matches_the_git_cli_at_the_crate_root() {
        let dir = Path::new(env!("CARGO_MANIFEST_DIR"));
        let Some(expected) = git_ls_files(dir) else {
            return;
        };
        let actual = collect_files_with_git(dir).expect("git2 should discover the repository");
        assert_eq!(actual, expected);
    }

    #[test]
    fn collect_files_with_git_matches_the_git_cli_for_a_subdirectory() {
        let dir = Path::new(env!("CARGO_MANIFEST_DIR")).join("src");
        let Some(expected) = git_ls_files(&dir) else {
            return;
        };
        let actual = collect_files_with_git(&dir).expect("git2 should discover the repository");
        assert_eq!(actual, expected);
    }

    #[test]
    fn is_git_workspace_root_matches_the_git_cli() {
        let manifest_dir = Path::new(env!("CARGO_MANIFEST_DIR"));
        assert!(!is_git_workspace_root(manifest_dir));
        let Some(toplevel) = git_ls_files(manifest_dir).map(|_| ()).and(
            Command::new("git")
                .args(["rev-parse", "--show-toplevel"])
                .current_dir(manifest_dir)
                .output()
                .ok(),
        ) else {
            return;
        };
        if !toplevel.status.success() {
            return;
        }
        let toplevel_dir = PathBuf::from(String::from_utf8_lossy(&toplevel.stdout).trim());
        assert!(is_git_workspace_root(&toplevel_dir));
    }
}
